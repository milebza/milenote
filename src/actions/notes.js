import uuid from 'uuid'
import database from '../firebase/firebase'

// components calls action generator
// action generator returns function
// component dispatches function
// function runs (has the ability to dispatch other actions and do whatever it wants)

// ADD_NOTE
export const addNote = (note) => ({
  type: 'ADD_NOTE',
  note
})

export const startAddNote = (noteData = {}) => {
  // this function only works due to thunk
  // it gets called internally by redux
  return (dispatch) => {
    // default data
    const {
      title = '',
      content = '',
      date = 0
    } = noteData
    const note = {title, content, date}

    return database.ref('notes').push(note).then((ref) => {
      dispatch(addNote({
        id: ref.key,
        ...note
      }))
    })
  }
}

// REMOVE_NOTE
export const removeNote = ({ id } = {}) => ({
  type: 'REMOVE_NOTE',
  id
})

// EDIT_NOTE
export const editNote = (id, updates) => ({
  type: 'EDIT_NOTE',
  id,
  updates
})
