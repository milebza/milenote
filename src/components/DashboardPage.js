import React from 'react'
import NoteList from './NoteList'
import NoteListFilters from './NoteListFilters'

const DashboardPage = () => (
  <div>
    <NoteListFilters />
    <NoteList />
  </div>
)

export default DashboardPage
