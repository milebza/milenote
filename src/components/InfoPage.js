import React from 'react'

const InfoPage = () => (
  <div className="form">
    <p className="text-center">Coded and designed by Milena Baeza</p>  
  </div>
)

export default InfoPage
