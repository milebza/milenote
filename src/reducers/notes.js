// Notes Reducer
const NotesReducerDefaultState = []

export default (state = NotesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_NOTE':
      return [
        ...state,
        action.note
      ]
    case 'REMOVE_NOTE':
        return state.filter(({ id }) => (id !== action.id))
    case 'EDIT_NOTE':
      return state.map((note) => {
        if (note.id === action.id) {
          return {
            ...note,
            ...action.updates
          }
        }
        return note
      })
    case 'SET_NOTES':
        return action.notes
    default:
      return state
  }
}
