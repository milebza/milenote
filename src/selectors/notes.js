import moment from 'moment'

// Get visible Notes
export default (notes, {text, startDate, endDate}) => {
  return notes.filter((note) => {
    const date = moment(note.date)
    const startDateMatch = startDate ? startDate.isSameOrBefore(date, 'day') : true
    const endDateMatch = endDate ? endDate.isSameOrAfter(date, 'day') : true
    const textMatch = note.title.toLowerCase().includes(text.toLowerCase()) || note.content.toLowerCase().includes(text.toLowerCase())

    return startDateMatch && endDateMatch && textMatch
  }).sort((a, b) => {
    // 1 -> b will go first
    return a.date < b.date ? 1 : -1
  })
}
