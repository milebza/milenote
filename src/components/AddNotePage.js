import React from 'react'
import { connect } from 'react-redux'
import { startAddNote } from '../actions/notes'
import NoteForm from './NoteForm'

export class AddNotePage extends React.Component {
  onSubmit = (note) => {
    this.props.startAddNote(note)
    this.props.history.push('/')
  }
  render() {
    return (
      <div>
        <NoteForm
          onSubmit={this.onSubmit}
        />
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddNote: (note) => dispatch(startAddNote(note))
})

export default connect(undefined, mapDispatchToProps)(AddNotePage)
