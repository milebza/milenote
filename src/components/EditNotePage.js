import React from 'react'
import { connect } from 'react-redux'
import { startEditNote, startRemoveNote } from '../actions/notes'
import Note from './Note'
import NoteForm from './NoteForm'

export class EditNotePage extends React.Component {
  state = {
    editView: false
  }
  onSubmit = (note) => {
    this.props.startEditNote(this.props.note.id, note)
    this.props.history.push('/')
  }
  onRemove = () => {
    this.props.startRemoveNote({ id: this.props.note.id })
    this.props.history.push('/')
  }
  handleEditView = () => {
    this.setState((prevState) => ({ editView: true }))
  }
  render() {
    return (
      <div className="relative">
        <button className="btn-link btn--left"
          onClick={this.onRemove}><span className="icon icon-bin"></span> Remove</button>

        {
          this.state.editView ?
          (
            <NoteForm
              note={this.props.note}
              onSubmit={this.onSubmit} />
          ) :
          (
            <div>
              <button className="btn-link form__btn form__btn-edit"
                onClick={this.handleEditView}><span className="icon icon-pencil"></span> Edit</button>
              <Note {...this.props.note} />
            </div>
          )
        }
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({ note: state.notes.find((note) => note.id === props.match.params.id) })

const mapDispatchToProps = (dispatch) => ({
  startEditNote: (id, note) => dispatch(startEditNote(id, note)),
  startRemoveNote: (data) => dispatch(startRemoveNote(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditNotePage)
