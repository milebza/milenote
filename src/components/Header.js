import React from 'react'
import { NavLink } from 'react-router-dom'
import { connect } from 'react-redux'
import { startLogout } from '../actions/auth'

export const Header = ({ startLogout }) => (
  <header className="header">
    <div className="flex header__content-wrapper">
      <h1 className="header__title">Your notes</h1>
      <button className="header__logout-btn" onClick={startLogout}>Logout</button>
    </div>
    <nav className="row header__nav">
      <NavLink to="/dashboard" activeClassName="header__link--active" className="col col-xs-4 header__link"><span className="icon icon-list2"></span></NavLink>
      <NavLink to="/add" activeClassName="header__link--active" className="col col-xs-4 header__link"><span className="icon icon-file-text2"></span></NavLink>
      <NavLink to="/info" activeClassName="header__link--active" className="col col-xs-4 header__link"><span className="icon icon-info"></span></NavLink>
    </nav>
  </header>
)

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
})

export default connect(undefined, mapDispatchToProps)(Header)
