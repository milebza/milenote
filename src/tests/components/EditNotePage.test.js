import React from 'react'
import { shallow } from 'enzyme'
import { EditNotePage } from '../../components/EditNotePage'
import notes from '../fixtures/notes'

let note, editNote, removeNote, history, wrapper

beforeEach(() => {
  note = notes[0]
  editNote = jest.fn()
  removeNote = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(<EditNotePage note={notes[0]} history={history} editNote={editNote} removeNote={removeNote} />)
})

test('should render EditNotePage correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle removeNote', () => {
  wrapper.find('button').at(0).simulate('click')
  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(removeNote).toHaveBeenLastCalledWith({id: notes[0].id})
})

test('should handle change to edit view', () => {
  expect(wrapper).toMatchSnapshot()
  wrapper.find('button').at(1).simulate('click')
  expect(wrapper).toMatchSnapshot()
})

test('should handle editNote', () => {
  wrapper.find('button').at(1).simulate('click')
  wrapper.find('NoteForm').prop('onSubmit')(notes[0])
  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(editNote).toHaveBeenLastCalledWith(notes[0].id, notes[0])
})
