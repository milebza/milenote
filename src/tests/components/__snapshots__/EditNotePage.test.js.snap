// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle change to edit view 1`] = `
<div
  className="relative"
>
  <button
    className="btn-link btn--left"
    onClick={[Function]}
  >
    <span
      className="icon icon-bin"
    />
     Remove
  </button>
  <div>
    <button
      className="btn-link form__btn form__btn-edit"
      onClick={[Function]}
    >
      <span
        className="icon icon-pencil"
      />
       Edit
    </button>
    <Note
      content="Dear friend"
      date={0}
      id="1"
      title="Letter"
    />
  </div>
</div>
`;

exports[`should handle change to edit view 2`] = `
<div
  className="relative"
>
  <button
    className="btn-link btn--left"
    onClick={[Function]}
  >
    <span
      className="icon icon-bin"
    />
     Remove
  </button>
  <NoteForm
    note={
      Object {
        "content": "Dear friend",
        "date": 0,
        "id": "1",
        "title": "Letter",
      }
    }
    onSubmit={[Function]}
  />
</div>
`;

exports[`should render EditNotePage correctly 1`] = `
<div
  className="relative"
>
  <button
    className="btn-link btn--left"
    onClick={[Function]}
  >
    <span
      className="icon icon-bin"
    />
     Remove
  </button>
  <div>
    <button
      className="btn-link form__btn form__btn-edit"
      onClick={[Function]}
    >
      <span
        className="icon icon-pencil"
      />
       Edit
    </button>
    <Note
      content="Dear friend"
      date={0}
      id="1"
      title="Letter"
    />
  </div>
</div>
`;
